arduino-ide-extension/src/browser/boards/boards-service-provider.ts

```diff
@@ -1,37 +1,137 @@
-import { injectable, inject } from '@theia/core/shared/inversify';
-import { Emitter } from '@theia/core/lib/common/event';
-import { ILogger } from '@theia/core/lib/common/logger';
+import { FrontendApplicationContribution } from '@theia/core/lib/browser/frontend-application';
+import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
 import {
   Command,
   CommandContribution,
   CommandRegistry,
   CommandService,
 } from '@theia/core/lib/common/command';
+import type { Disposable } from '@theia/core/lib/common/disposable';
+import { Emitter } from '@theia/core/lib/common/event';
+import { ILogger } from '@theia/core/lib/common/logger';
 import { MessageService } from '@theia/core/lib/common/message-service';
-import { FrontendApplicationContribution } from '@theia/core/lib/browser/frontend-application';
-import { RecursiveRequired } from '../../common/types';
+import { nls } from '@theia/core/lib/common/nls';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import type { Mutable } from '@theia/core/lib/common/types';
+import { inject, injectable, optional } from '@theia/core/shared/inversify';
 import {
-  Port,
-  Board,
-  BoardsService,
+  OutputChannel,
+  OutputChannelManager,
+} from '@theia/output/lib/browser/output-channel';
+import {
+  BoardIdentifier,
+  boardIdentifierEquals,
+  BoardsConfig,
+  BoardsConfigChangeEvent,
   BoardsPackage,
-  AttachedBoardsChangeEvent,
-  BoardWithPackage,
+  BoardsService,
   BoardUserField,
-  AvailablePorts,
+  BoardWithPackage,
+  DetectedPorts,
+  emptyBoardsConfig,
+  isBoardIdentifier,
+  isPortIdentifier,
+  Port,
+  PortIdentifier,
+  portIdentifierEquals,
+  serializePlatformIdentifier,
 } from '../../common/protocol';
-import { BoardsConfig } from './boards-config';
-import { naturalCompare } from '../../common/utils';
-import { NotificationCenter } from '../notification-center';
-import { StorageWrapper } from '../storage-wrapper';
-import { nls } from '@theia/core/lib/common';
-import { Deferred } from '@theia/core/lib/common/promise-util';
-import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
-import { Unknown } from '../../common/nls';
+import {
+  BoardList,
+  BoardListHistory,
+  createBoardList,
+  isBoardListHistory,
+} from '../../common/protocol/board-list';
+import { Defined } from '../../common/types';
 import {
   StartupTask,
   StartupTaskProvider,
 } from '../../electron-common/startup-task';
+import { NotificationCenter } from '../notification-center';
+import { StorageWrapper } from '../storage-wrapper';
+import { BoardsDataStore } from './boards-data-store';
+
+const boardListHistoryStorageKey = 'arduino-ide:boardListHistory';
+const selectedPortStorageKey = 'arduino-ide:selectedPort';
+const selectedBoardStorageKey = 'arduino-ide:selectedBoard';
+
+type UpdateBoardsConfigReason =
```

Comment:
I like the design, the descrete "reasons" for updating the `boardsConfig`. One thing that came to mind when I saw this type; `UpdateBoardsConfigReason`, was: "to which reason/category would port update due to upload belong?"

Looking at the logic as a whole I see we are calling `updatePort` directly from `upload-sketch.ts`. I don't think there is anything problematic about that, but it feels slightly disjointed to me.

I would expect us to rely on a single method for all updates to the `boardConfig` (external or internal to the class), and in this case have a reason like `'upload'`, calling a method that asserts that reason somewhere.

The result I have in my mind is the ability to see at a glance in the module all the reasons for which changes are made programtically to selected board or port, plus greater ease when searching the codebase for places where we invoke these changes.

Reply:
> was: "to which reason/category would port update due to upload belong?"
It does not belong to any categories as it is unnecessary. If the upload changes the port, it's not a user interaction.

The board+port setting (aka `BoardsConfig`) can change (both or either of them) in four different ways:
 - the user tweaks the dialog and OKs it,
 - the user clicks any of the board list items where both board+port are defined,
 - the user creates a new sketch (and the new window inherits the config),
 - the user starts IDE2.

I am unaware of any other user interaction where the user can modify one or both. The user can change the board or port from the (`Tools`) menu, not an entire `BoardsConfig`.

IDE2 can change the board:
 - when a board is selected, but the platform is not installed, and the user installs it.

IDE2 can change the port:
 - when it changes during the upload.

> I would expect us to rely on a single method for all updates to the `boardConfig` (external or internal to the class), and in this case have a reason like `'upload'`, calling a method that asserts that reason somewhere.

This is precisely what I want to avoid: having one magic method that does everything.

> for which changes are made programmatically

Please see above" `updatePort` and `updateBoard` were made to handle programmatic changes. `updateConfig` is the API to change the `BoardsConfig` as a side effect of user interaction.

Reply:
> instead, something like:
> 
> ```
> updateBoardsConfig(payload, 'upload')
> ```

I see what you mean. Thanks for explaining. OK, it makes sense. Let's do it.

Reply:
> OK, it makes sense. Let's do it.

Sure, we can if you agree there are merits to the idea! Although I don't feel we **have** to change it. Now that I understand a bit more the divisions, I think the reasoning behind current design is sound.

> what is here feels like an appropriate solution

^ maybe it's already clear, but with this vague comment, by: "_what is here_", I meant what we already have in the PR, that it's appropriate anyway

Reply:
I rewrote it and provided one entry to update the board, the port, or both.

I am resolving this thread now. Please open a new one if you need to. Thank you!